dashboard/utils.py

```diff
@@ -210,6 +184,42 @@ def has_final_grade(self, edx_course_key):
         """
         return self.final_grade_qset.for_course_run_key(edx_course_key).exists()
 
+    def has_paid(self, edx_course_key):
+        """
+        Returns true if user paid for a course run.
+
+        Args:
+            edx_course_key (str): an edX course run key
+
+        Returns:
+            bool: whether the user is paid
+        """
+        # financial aid programs need to have a paid entry for the course
+        if self.financial_aid_available:
+            return edx_course_key in self.paid_course_keys
+
+        # normal programs need to have paind_on_edx in the final grades or a verified enrollment
+        if self.has_final_grade(edx_course_key):
+            return self.has_paid_on_edx_final_grade(edx_course_key)
+        return self.has_verified_enrollment(edx_course_key)
+
+    def has_paid_for_any_in_program(self):
+        """
+        Returns true if a user has paid for any course run in the program
+        """
+        return any(self.has_paid(edx_course_key) for edx_course_key in self.edx_course_keys)
+
+    def has_paid_on_edx_final_grade(self, edx_course_key):
+        """
+        Checks if there is a a frozen final grade and the user paid for it on edx.
+
+        Args:
+            edx_course_key (str): an edX course run key
+        Returns:
+            bool: whether a frozen final grade with the flag paid on edx exists
+        """
+        return self.final_grade_qset.paid_on_edx().for_course_run_key(edx_course_key).exists()
+
     def has_paid_final_grade(self, edx_course_key):
         """
         Checks if there is a a frozen final grade and the user paid for it.
```

Comment:
not related to this PR, but might as well fix it here: Checks if there is a ~~a~~ frozen final grade and the user paid for it.